### ISSUE TITLE
🚀 make whitespace changes visible as such

### ISSUE THREAD

**Issue by Kr1ss-XD** [dandavison/delta#522]
Hi, at first glance I couldn't find any open or closed issues on this topic. Please excuse me if I missed one.

Recently, while viewing a particular diff, I noticed that every indented line was highlighted as a difference, even though the code itself was not changed, other than being indented by _spaces_ vs. _tabs_. (PKGBUILD's are interpreted by bash, so indentation doesn't matter).

![delta_whitespace1](https://user-images.githubusercontent.com/28707521/107787047-e1512400-6d4e-11eb-8d84-5a0246174343.png)

Therefore, I think it might be useful to have an option which allows displaying whitespace characters using one or more user-defined glyphs, and/or applying the `delta.{plus,minus}-emph-style` settings to whitespace changes.

In the case that the characters to display tabs were be defined as `'‣·'` and tab size as `4`, I would imagine something like :

![delta_whitespace2](https://user-images.githubusercontent.com/28707521/107786999-d39b9e80-6d4e-11eb-8255-c87eb53eebdf.png)

**Comment by dandavison** [comment]
Hi @Kr1ss-XD can I just check -- passing `-w` (`--ignore-all-space`) to git diff/show/etc  works with delta already. Are you suggesting the new delta functionality would be worth adding, in addition to what we have with `-w`?

**Comment by Kr1ss-XD** [comment]
Yes, that's correct. I would like to be able to see whitespace changes, but also to distinguish them from other differences on a glance.

**Comment by dandavison** [comment]
I see. It almost feels like there could be a git option for rendering tabs as visible characters doesn't it? But there doesn't seem to be one. Just playing around with possibilities, this seems quite nice:

```gitconfig
[core]
    pager = sed 's/\t/├──┤/g' | delta
```

So is this as simple as allowing the user to supply a string to substitute tabs? Here's a quick branch implementing that, do you want to try it out? https://github.com/dandavison/delta/pull/523 E.g.

```
git -c delta.tab-string='├──┤' diff
```

<table><tr><td><img width=800px src="https://user-images.githubusercontent.com/52205/107811466-30418e00-6d3c-11eb-9836-81596caac7f6.png" alt="image" /></td></tr></table>

**Comment by Kr1ss-XD** [comment]
Nice, I guess that's exactly what I was looking for.

I'll check out that branch and build it. Thanks a lot !

**Comment by Kr1ss-XD** [comment]
It's working fine @dandavison :+1:

**Comment by dandavison** [comment]
Great. One question is do we consider the breaking change of getting rid of the `tabs` option, since if we go with the new branch it's redundant since any number of spaces can be provided explicitly. I guess it can be marked as deprecated (but we are not yet at 1.0 so I suppose the point of 0.x is you're allowed to break backwards compatibility while setting down.)

**Comment by Kr1ss-XD** [comment]
I'd say there are - at least - two ways to handle this.

1. deprecate `delta.tabs` in favor of `delta.tab-string` (and break backwards compatibility), or
2. maybe keep `tabs`, and let `tab-string` accept 1, 2 or 3 characters. If one character is given, repeat it `tabs` times. If there are two chars, print the first one and repeat the 2nd `tabs`-1 times. If `tab_strings` contains three characters, print the 1st one, print the second one `tabs`-2 times, then print the 3rd character.

>I suppose the point of 0.x is you're allowed to break backwards compatibility while setting down.

That's my understanding, too. Though I'd say it's preferable to prevent breaking changes if possible.

**Comment by th1000s** [comment]
Speaking of tabs, when replacing them `delta` currently uses a fixed number of spaces. However these replacements must stop at a multiple of the tab width, counted from the start of the line:

    1\tX                                   1       X                               1‣······X
    12\tX         which looks like         12      X     and delta then might      12‣·····X
    123\tX        this with a tab-         123     X     render them like this:    123‣····X
    1234\tX       width of 8               1234    X                               1234‣···X
    1234567\tX                             1234567 X                               1234567‣X

Also maybe use a more muted color than white, e.g. gray or blue as `vim` does with `:set list`.

**Comment by robinmoussu** [comment]
If possible, I think it would be great to make clearer that lines that contains only whitespace change don’t introduce other modifications.

<img width="456" alt="current" src="https://user-images.githubusercontent.com/6200547/117118022-8567cb80-ad90-11eb-9c55-e7c9f9035627.png">

In this example, it doesn’t seems like the first line is really an addition, because of the whitespace changes on the next lines. I think the easiest change would be to use the light green background for lines that are entirely new. The same logic can be used for lines that where completely removed.

<img width="456" alt="proposed" src="https://user-images.githubusercontent.com/6200547/117118410-04f59a80-ad91-11eb-999d-c2eea342adc6.png">

**Comment by th1000s** [comment]
@robinmoussu also see PR #13 which among other things discusses that single changed words are emphasized more than an entire added or removed line.

**Comment by alfredbez** [comment]
Just for reference: Git also [has an option](https://stackoverflow.com/a/58037244/2123108) to show invisible or non-printable characters via `--ws-error-highlight=all`.

**Comment by mentalisttraceur** [comment]
@dandavison I think the strongest reason was the quote (emphasis added) "`delta` does not use `bat` when it is running (*it does use some code from the excellent `bat` project*, but users don't need to be aware of this)" on [your page about `delta`'s use of pager environment variables](https://dandavison.github.io/delta/environment-variables.html?highlight=bat#pager-environment-variables), compounded by other mentions of `bat` in the context of `delta` themes such as [here](https://dandavison.github.io/delta/supported-languages-and-themes.html?highlight=bat#supported-languages-and-themes) and [here](https://dandavison.github.io/delta/custom-themes.html). (And maybe I've seen other people say it too, I can't remember.)

All together, that gave me the impression that there was significant code reuse, and my first thought would be to try to reuse it for any heavy lifting of syntax highlighting the text.

**Comment by dandavison** [comment]
Right, it's funny, I added those sentences to try to dispel the belief that delta used code from bat for syntax highlighting, but I can see it's confusing especially if people come with a prior belief that it does. Delta uses bat code for things like launching the pager process, and it uses the color themes and language definitions that bat uses. But the low-level details of syntax highlighting are implemented independently in bat and delta, both using the same underlying library (syntect).

**Comment by mentalisttraceur** [comment]
Gotcha, thanks for clarifying!

If I think of any better wording or place for more precisely/unambiguously mitigating that misunderstanding I'll let you know!

In any case, I would love a `--show-all` option, and I think `bat` is a good example of it done well.

**Comment by dandavison** [comment]
@mentalisttraceur, I agree, I use `bat -A` a lot.

I haven't thought too hard about how helpful this is but FWIW here's a conversation with GPT4 about this issue.

-------

> Can you summarize the following git-delta issue and suggest next steps?
> [paste entire text copied from browser including title]

The git-delta issue #522 discusses the need for a feature to make whitespace changes visible and distinguishable from other changes in a diff. The main problem raised is that when viewing a diff, indented lines are highlighted as differences, even though the code is not changed, but only indented with spaces or tabs. The proposed solution is to have an option to display whitespace characters using user-defined glyphs and/or applying specific delta styles to whitespace changes.

Throughout the conversation, there are various suggestions and implementations shared. The initial idea is to allow the user to supply a string to substitute tabs. However, it evolves to include discussions about how to handle the deprecated delta.tabs in favor of delta.tab-string and whether to break backward compatibility.

The conversation also covers topics such as:

- Tab alignment and the possibility of using a vim-style configuration for tab replacement characters.
- Showing invisible or non-printable characters with Git's --ws-error-highlight=all option.
- The possibility of adding a --show-all option similar to the bat tool.

Next steps:

1. Assess the feasibility and potential impact of the suggested implementations and decide on the most suitable approach.
2. Implement the chosen solution, ensuring that it addresses the issue raised and is compatible with existing features.
3. Test the implemented solution thoroughly to ensure it works as expected and does not introduce new issues.
4. Update documentation and provide examples of how to use the new feature.
Close the issue after successful implementation and testing.

> Can you give a specific suggestion for a next step as follows:
> 1. The proposed change to delta's behavior and user experience
> 2. The proposed implementation

1. Proposed change to delta's behavior and user experience:

  - Introduce a --show-all option that, when enabled, will display all invisible characters in the diff, such as spaces and tabs. This option will help users to quickly identify and distinguish between whitespace changes and other changes in the code. The feature should be inspired by the similar functionality in the bat tool, ensuring that it is familiar and intuitive for users.

2. Proposed implementation:

  - Add a new command-line flag --show-all to the list of options accepted by delta.
  - In the processing stage where delta is handling the diff, check if the --show-all flag is enabled. If so, replace the invisible characters (e.g., spaces and tabs) with their corresponding visible glyphs (e.g., '·' for space and '‣' for tab).
  - Ensure that the new feature is compatible with other delta options such as --ignore-all-space and works seamlessly with delta's styles and themes.
  - Add tests to verify that the new feature works as expected and does not introduce new issues or conflicts with existing functionality.
  - Update the documentation and help messages to include the new --show-all option, providing examples of its usage and explaining its purpose.

> Is the idea of the `--show-all` option to explain to the user _differences_ in whitespace between the old and new version of the file, or is it simply to show which whitespace characters are in use even when they haven't changed?

Apologies for any confusion in my previous response. The idea of the --show-all option is to display all whitespace characters in use, regardless of whether they have changed or not. This will help users visualize the actual structure of the code, including spaces and tabs, making it easier for them to understand the formatting and indentation. It does not specifically highlight the differences in whitespace between the old and new versions of the file. However, by making the whitespace characters visible, users can more easily identify and compare the changes in whitespace themselves.

> How do you propose that delta should display differences in whitespace characters between the old and new version?

Delta could use a combination of colors and symbols to indicate the differences in whitespace characters. For example, it could use different colors for changed spaces and tabs, and a separate color for unchanged whitespace characters.

Here is an ASCII representation of how it could look:

```
Old version:
if (condition) {
····doSomething();
}

New version:
if (condition) {
‣‣‣‣doSomething();
}
```

In this example:

- · represents a space character from the old version
- ‣ represents a tab character from the new version
Delta could use a light red background color for the removed spaces (represented by ·) and a light green background color for the added tabs (represented by ‣). This would make it clear that spaces were removed and replaced with tabs.

Please note that this is a text-based representation, and the actual implementation in delta would involve colors and custom symbols to create a more visually appealing and informative display.

**Comment by mentalisttraceur** [comment]
Sounds like I convinced ChatGPT4 😄.

I appreciate that you publicly shared that - I hadn't thought to use AI as a thinking aid like that, seems useful!

I basically agree with its recommendations - obviously to be precise, we probably don't want a full tab represented by just *one* single-width character (I'm tempted to say "that's what it meant", in the sense that if a human said this, I'd suppose that's probably what they meant).

As for coloring, I imagine the coloring of the whitespace characters to be like syntax highlighting now: if style includes `syntax`, the special character stand-ins would have the same foreground/text colors on all lines (maybe just copy how `bat -A` colors them), and the background color would indicate removed/added/moved. (Of course if someone customizes the style to have a foreground color override the syntax coloring, that overrides.)

**Comment by Freed-Wu** [comment]
> passing -w (--ignore-all-space) to git diff/show/etc works with delta already.

Can we directly pass the option to delta, not diff? such as `delta --ignore-all-space a.txt b.txt`, not `diff -u --ignore-all-space a.txt b.txt | delta`

**Comment by dandavison** [comment]
> Can we directly pass the option to delta, not diff? such as `delta --ignore-all-space a.txt b.txt`, not `diff -u --ignore-all-space a.txt b.txt | delta`

Yes we can. See

```
  -@, --diff-args <STRING>   Extra arguments to pass to `git diff` when using delta to diff two files
```

<img width="283" height="432" alt="Image" src="https://github.com/user-attachments/assets/04ee0b29-b450-4c73-bb20-63224ddbc4a9" />
